Feature/compost block

Fireruner opened the pull request:
# Changes proposed in this pull request:
- Added the Compost Item, an Item that behaves like BoneMeal to the plants and cannot be crafted nor used as a dye.
- Adde the Barrel Block, a block that takes biolical items (can be set on configurations) and process them. After 2 days (average) the block will emit particles to show that it finished, and will drop 6 Compost Items when the player interacts whit it.

Review please

Fireruner committed:
Compost Block

Not working at the moment. Crashes on launch

Fireruner committed:
Compost Block Model

Added the model for the compost block.

Fireruner committed:
Composter Block Model Fixes and Very Basic Functionality

Fixed some problems with the model, and added the capability of changing its state with right click.

Fireruner committed:
Barrel Block Almost Ready

The basic functionality works, but it sometimes "forgets" that it has the items it had. Have to work on that. For now it drops 1 bone from each 64 mats (rotten flesh counts as double) in like 1-2 mins (will increase time and drop)

Fireruner committed:
Barrel Block Little Fixes

Little fixes here and there. Still "eating" items

Fireruner committed:
BarrelBlock Funtionality

Functionality of the barrel block working. Pending of testing and configuring (time needed, drop amounts, etc).

Fireruner committed:
Barrel Block New Model

Added a model for the barrel while its working (the same barrel but with a lid on it)

Fireruner committed:
Barrel Block 

Added particles to the finish state and directions for the block. Pending of final details.

Fireruner committed:
Barrel State Orientation Fix

Fixed a problem where the barrel orientation will change randomly in servers. Test pending

Fireruner committed:
Trying to fix sinc problems

Looks like the client isn´t getting the state of the barrel quite right. It gets the orientation right, though

Fireruner committed:
Merge branch 'version/1.12' into feature/compostBlock

Raycoms committed:
comments to fix this

Fireruner committed:
Update problems with the barrel

Trying to get the barrel updated every time the player uses it. For now you have to force the update modifying blocks arround it.

Fireruner committed:
Trying to get the barrel to update

Fireruner committed:
Still refusing to update

Fireruner committed:
Merge branch 'version/1.12' into feature/compostBlock

Fireruner committed:
Barrel Updating!

The barrel updates fine in singleplayer and works as intended. Minor adjustments pending.

Fireruner committed:
AI interface

Added the signature for the methods to be used as interface with the AI

Fireruner committed:
Barrel sync and configs

Logic now gets executed on server and the barrel gets synced in server and client. Added a configuration to set, unset items that can be processed by the barrel

Fireruner committed:
Merge branch 'version/1.12' into feature/compostBlock

Fireruner committed:
Fixing gradle mistake

Fireruner committed:
Composter finished and started compost

The BarrelBlock is done and working (pending from further testing), and the compost item was included in the items and its funcionallity done, pending from graphics.

Fireruner committed:
Barrel Block and Compost Item finished

They are both working and pending from testing.

Fireruner committed:
Finished!

I hope

Fireruner committed:
Sonar

Fixing Sonar issues

Fireruner committed:
Sonar II

Still fixing

Fireruner committed:
Sonar III

Fireruner committed:
Sonar IV

LDTTeam-Buildsystem commented on src/main/java/com/minecolonies/coremod/tileentities/TileEntityBarrel.java:
```diff
@@ -0,0 +1,240 @@
+        if(this.done)
+        {
+            //If the barrel is done, we spawn particles imitating "bad smell"
+            ((WorldServer)worldIn).spawnParticle(
```
### Errors:
- Casting worldIn to WorldServer may produce java.lang.ClassCastException

Fireruner replied:
Will never produce the exception as it never gets called if world.isRemote is true, wich means the world is a server. Not need to check if it's an instance of WorldServer and it would be too expensive.

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34035 |  0  |  3  |  -3  |
| Errors |  594  |  0  |  0  |  0   |
##### More Information:

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34017 |  0  | 18  | -18  |
| Errors |  594  |  0  |  0  |  0   |
##### More Information:

Fireruner committed:
Sonar V

Fireruner committed:
Sonar VI

Raycoms commented on src/main/java/com/minecolonies/coremod/blocks/BlockBarrel.java:
```diff
@@ -74,163 +88,230 @@ private void initBlock()
+    @Override
+    public IBlockState getStateForPlacement(
+      @NotNull final World world, @NotNull final BlockPos pos, @NotNull final EnumFacing facing, final float hitX, final float hitY,
+                                            final float hitZ, final int meta, @NotNull final EntityLivingBase placer, final EnumHand hand) {
```
formatting is off here

Raycoms commented on src/main/java/com/minecolonies/coremod/blocks/types/BarrelType.java:
```diff
@@ -0,0 +1,90 @@
+        this(metaIn, nameIn, nameIn, mapColorIn);
+    }
+
+    BarrelType(final int metaIn, final String nameIn, final String unlocalizedNameIn, final MapColor mapColorIn)
```
javadoc for all public methods and fields please

Raycoms commented on src/main/java/com/minecolonies/coremod/items/ItemCompost.java:
```diff
@@ -5,24 +5,25 @@
-    /**
-     * Max stacksize of the item.
-     */
     private static final int MAX_STACK_SIZE = 64;
```
we have a constant for this somewhere already.

Raycoms commented on src/main/java/com/minecolonies/coremod/items/ItemCompost.java:
```diff
@@ -5,24 +5,25 @@
-    /**
-     * Compost constructor, Set max stack size to 64 like all other items.
-     */
     public ItemCompost()
```
Please leave doc,

Raycoms commented on src/main/java/com/minecolonies/coremod/tileentities/TileEntityBarrel.java:
```diff
@@ -0,0 +1,240 @@
+{
+    private boolean                                 done           = false;
+    private int                                     items          = 0;
+    private int                                     timer          = 0;
```
jdoc for public methods and for all fields please.
Comments please  in jdoc and not with // on next line

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34006 |  0  |  3  |  -3  |
| Errors |  594  |  0  |  0  |  0   |
##### More Information:

Fireruner commented on src/main/java/com/minecolonies/coremod/tileentities/TileEntityBarrel.java:
```diff
@@ -0,0 +1,240 @@
+        if(this.done)
+        {
+            //If the barrel is done, we spawn particles imitating "bad smell"
+            ((WorldServer)worldIn).spawnParticle(
```
Will never produce the exception as it never gets called if world.isRemote is true, wich means the world is a server. Not need to check if it's an instance of WorldServer and it would be too expensive.

Fireruner committed:
Comments I

Fixing the formatting and adding JavaDoc where needed

Raycoms commented:
:shipit: looks good for me now!

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/Minecolonies/minecolonies/pull-request/2605/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=Minecolonies/minecolonies)

marchermans reviewed (dismissed):
Okey all good. Just one question: Should we not add Support for OreDictionary?

Fireruner committed:
Added OreDictionay support

All the item names and tags are defined in the same config entry. Added OreDictionary registries to the vanilla items where needed.

Fireruner committed:
Fixing minor bug

Wasn't checking the hole array of ids

Raycoms commented:
:shipit: looks good
